include/deal.II/base/floating_point_comparator.h

```diff
@@ -0,0 +1,272 @@
+ * comparator class within a std::map<> of the given arrays. Note that this
+ * comparison operator does not satisfy all the mathematical properties one
+ * usually wants to have (consider e.g. the numbers a=0, b=0.1, c=0.2 with
+ * tolerance 0.15; the operator gives a<c, but neither a<b? nor b<c? is
+ * satisfied). This is not a problem in the use cases for this class, but be
+ * careful when using it in other contexts.
+ */
+template <typename Number>
+struct FloatingPointComparator
+{
+  using ScalarNumber =
+    typename dealii::internal::VectorizedArrayTrait<Number>::value_type;
+  static constexpr std::size_t width =
+    dealii::internal::VectorizedArrayTrait<Number>::width;
+
+  /**
+   * Constructor.
+   *
+   * TODO: What is scaling?
+   */
+  FloatingPointComparator(
+    const ScalarNumber       scaling,
+    const std::bitset<width> mask = std::bitset<width>().flip());
+
+  /**
+   * Compare two vectors of numbers (not necessarily of the same length)
+   */
+  bool
+  operator()(const std::vector<ScalarNumber> &v1,
+             const std::vector<ScalarNumber> &v2) const;
+
+  /**
+   * Compare two vectorized arrays (stored as tensors to avoid alignment
+   * issues).
+   */
+  bool
+  operator()(const Tensor<1, width, ScalarNumber> &t1,
+             const Tensor<1, width, ScalarNumber> &t2) const;
+
+  /**
+   * Compare two rank-1 tensors of vectorized arrays (stored as tensors to
+   * avoid alignment issues).
+   */
+  template <int dim>
+  bool
+  operator()(const Tensor<1, dim, Tensor<1, width, ScalarNumber>> &t1,
+             const Tensor<1, dim, Tensor<1, width, ScalarNumber>> &t2) const;
+
+  /**
+   * Compare two rank-2 tensors of vectorized arrays (stored as tensors to
+   * avoid alignment issues).
+   */
+  template <int dim>
+  bool
+  operator()(const Tensor<2, dim, Tensor<1, width, ScalarNumber>> &t1,
+             const Tensor<2, dim, Tensor<1, width, ScalarNumber>> &t2) const;
+
+  /**
+   * Compare two arrays of tensors.
+   */
+  template <int dim>
+  bool
+  operator()(const std::array<Tensor<2, dim, ScalarNumber>, dim + 1> &t1,
+             const std::array<Tensor<2, dim, ScalarNumber>, dim + 1> &t2) const;
+
+  /**
+   * Compare two tables.
+   */
+  template <typename T>
+  bool
+  operator()(const Table<2, T> &t1, const Table<2, T> &t2) const;
+
+  bool
+  operator()(const ScalarNumber &s1, const ScalarNumber &s2) const;
+
+  bool
+  operator()(const VectorizedArray<ScalarNumber, width> &v1,
+             const VectorizedArray<ScalarNumber, width> &v2) const;
+
+private:
+  ScalarNumber             tolerance;
+  const std::bitset<width> mask;
+};
+
+
+/* ------------------------------------------------------------------ */
+
+
+template <typename Number>
+FloatingPointComparator<Number>::FloatingPointComparator(
+  const ScalarNumber       scaling,
+  const std::bitset<width> mask)
+  : tolerance(scaling * std::numeric_limits<double>::epsilon() * 1024.)
```

Comment:
I find the definition of the tolerance a bit strange. Could we move this logic to the calling side? Especially the `double` part is a bit odd if I want to use it for `float`.

Reply:
We can certainly move setting the tolerance to the caller, in case we see the need for a different tolerance. The reason for having `double` tolerances has been done on purpose and based on many experiments as well as the nature of accumulation of roundoff in the routines subject to compression. You can understand this as an underlying assumption that the numbers originate from computations done in `double` precision, with the compression appended afterwards. The check for `float` then means that those need to be exactly equal without `float` roundoff error, as one ulp in `float` is more than `1e-13`. One could reason whether one should allow a coarser tolerance on `float`, but that would be need a case-by-case decision, as many assumptions build implicitly on the fact that `1e-7` is not yet zero, whereas `1e-12` is.

Reply:
@kronbichler What do you think about the second commit?